File: python/questions/decodeString.py
class Solution:
    def decodeString(self, s: str) -> str:
        ans = ""
        p = 0
        while p < len(s):
            oc = ord(s[p])
            if oc>=ord('a') and oc<=ord('z'):
                ans += s[p]
                p += 1
            else:
                rt = 0
                while s[p] != '[':
                    rt *= 10
                    rt += int(s[p])
                    p += 1
                lmr = 1
                sp = p + 1
                while lmr:
                    p += 1
                    if s[p] == '[':
                        lmr += 1
                    elif s[p] == ']':
                        lmr -= 1
                ss = self.decodeString(s[sp:p])
                ans += ss * rt
                p += 1
        return ans

File: cpp/questions/decodeString.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    string decodeString(string s) {
        std::string ans = "";
        int p = 0;
        while (p < s.size()) {
            if (s[p]>='a' && s[p]<='z') {
                ans.push_back(s[p]);
                ++p;
            } else {
                int rt = 0;
                while (s[p] != '[') {
                    rt *= 10;
                    rt += s[p]-'0';
                    ++p;
                }
                int sp = ++p;
                int lmr = 1;
                while (lmr) {
                    if (s[p] == '[') {
                        lmr += 1;
                    } else if (s[p] == ']') {
                        lmr -= 1;
                    }
                    ++p;
                }
                std::string ss = decodeString(s.substr(sp, p-1-sp));
                while (rt--) {
                    ans.append(ss);
                }
            }
        }
        return ans;
    }
};
